src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DisposableFieldsShouldBeDisposed.cs

```diff
@@ -41,172 +41,217 @@ public override void Initialize(AnalysisContext context)
             context.EnableConcurrentExecution();
             context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze);
 
-            context.RegisterCompilationStartAction(compilationContext =>
+            context.RegisterSymbolStartAction(OnSymbolStart, SymbolKind.NamedType);
+        }
+
+        private static void OnSymbolStart(SymbolStartAnalysisContext symbolStartContext)
+        {
+            if (!DisposeAnalysisHelper.TryGetOrCreate(symbolStartContext.Compilation, out DisposeAnalysisHelper? disposeAnalysisHelper))
             {
-                if (!DisposeAnalysisHelper.TryGetOrCreate(compilationContext.Compilation, out DisposeAnalysisHelper? disposeAnalysisHelper))
-                {
-                    return;
-                }
+                return;
+            }
 
-                var fieldDisposeValueMap = new ConcurrentDictionary<IFieldSymbol, /*disposed*/bool>();
-                void addOrUpdateFieldDisposedValue(IFieldSymbol field, bool disposed)
-                {
-                    Debug.Assert(!field.IsStatic);
-                    Debug.Assert(disposeAnalysisHelper!.IsDisposable(field.Type));
+            var fieldDisposeValueMap = PooledConcurrentDictionary<IFieldSymbol, /*disposed*/bool>.GetInstance();
+
+            var hasErrors = false;
+            symbolStartContext.RegisterOperationAction(_ => hasErrors = true, OperationKind.Invalid);
 
-                    fieldDisposeValueMap.AddOrUpdate(field,
-                        addValue: disposed,
-                        updateValueFactory: (f, currentValue) => currentValue || disposed);
-                };
+            // Disposable fields with initializer at declaration must be disposed.
+            symbolStartContext.RegisterOperationAction(OnFieldInitializer, OperationKind.FieldInitializer);
 
-                var hasErrors = false;
-                compilationContext.RegisterOperationAction(_ => hasErrors = true, OperationKind.Invalid);
+            // Instance fields initialized in constructor/method body with a locally created disposable object must be disposed.
+            symbolStartContext.RegisterOperationBlockStartAction(OnOperationBlockStart);
 
-                // Disposable fields with initializer at declaration must be disposed.
-                compilationContext.RegisterOperationAction(operationContext =>
+            // Report diagnostics at symbol end.
+            symbolStartContext.RegisterSymbolEndAction(OnSymbolEnd);
+
+            return;
```

Comment:
Do you need this return?

Reply:
It is generally recommended to have an explicit return statement separating the core logic and local functions, which are primarily implementation details. We do so even in roslyn repo.